### ISSUE TITLE
Broken Adorner Layer Invalidation or Clipping

### ISSUE THREAD

**Issue by robloo** [AvaloniaUI/Avalonia#10701]
**Describe the bug**

Just see the screen capture.

**To Reproduce**

AdornerLayer tab in the Control Catalog

**Expected behavior**

 1. I'm not sure if clipping to the page is enabled, but if so, it should clip by default
 2. Adorner lines should not be frozen and never refreshed
 3. Adorner lines should always draw over the slider
 4. Adorner lines really shouldn't have tearing and should always be straight

**Screenshots**

https://user-images.githubusercontent.com/17993847/226121183-cb584d8c-5423-42fa-9096-f5daec560e57.mp4

**Desktop (please complete the following information):**

- OS: Windows 10 Pro
- Version [e.g. 0.10.0-rc1 or 0.9.12]

**Additional context**

**Comment by HermanKirshin** [comment]
Any hints in which direction should I dig to fix this properly in Avalonia sources? It is close to become critical for our UI.

**Comment by timunie** [comment]
@HermanKirshin you can always request paid support / paid development in case something is critical for you.

**Comment by HermanKirshin** [comment]
I doubt that employer will allocate funds for third-party **regression**.
Anyway I found workaround for adorners that are in rather small area around control - control should be wrapped in transparent Border with ICustomHitTest implementation, negative margin (that should cover adorner area) and same value positive padding. Also with ClipToBounds=False. Maybe this can help someone too.

**Comment by billhenn** [comment]
We are also running into numerous adorner clipping/invalidation issues in Avalonia.  Here is a simple reproduction of several things (XAML code below):

![AdornerIssue](https://github.com/AvaloniaUI/Avalonia/assets/5760058/d1670f48-4009-49db-9a3e-848d0d3f649d)

1) The Red and Blue dots should be clipped within the `ScrollViewer` but they are not.  WPF does clip adorners in a `ScrollViewer`.

2) When you scroll down, the dot parts that shouldn't have been rendered outside the `ScrollViewer` remain there and don't go away.

3) If you resize the window (which I do at the end of the animation), it will cause the repaint to occur, erasing the artifacts.  But the red dot adorner now appears in its new location almost completely outside the `ScrollViewer`.  Only the very bottom of the red dot still in `ScrollViewer` bounds should be visible.

```
<Border Margin="0,50,0,0" BorderThickness="1" BorderBrush="Gray" HorizontalAlignment="Left">
	<ScrollViewer MaxHeight="120">
		<StackPanel Margin="20,0" Spacing="30">
			<Border Width="100" Height="80" BorderBrush="Gray" BorderThickness="3" Background="Silver" CornerRadius="30">
				<AdornerLayer.Adorner>
					<Canvas>
						<Ellipse Width="40" Height="40" Fill="Red">
							<Ellipse.RenderTransform>
								<TranslateTransform X="30" Y="-20" />
							</Ellipse.RenderTransform>
						</Ellipse>
					</Canvas>
				</AdornerLayer.Adorner>
			</Border>

			<Border Width="100" Height="80" BorderBrush="Gray" BorderThickness="3" Background="Silver" CornerRadius="30">
				<AdornerLayer.Adorner>
					<Canvas>
						<Ellipse Width="40" Height="40" Fill="Blue">
							<Ellipse.RenderTransform>
								<TranslateTransform X="30" Y="-20" />
							</Ellipse.RenderTransform>
						</Ellipse>
					</Canvas>
				</AdornerLayer.Adorner>
			</Border>
		</StackPanel>
	</ScrollViewer>
</Border>
```

I'm happy to assist looking into these issues or testing any fixes.

**Comment by billhenn** [comment]
After some debugging into the Avalonia source, I see a couple things...

## IsClipEnabled Property Default Issue

[This line](https://github.com/AvaloniaUI/Avalonia/blob/1f919606b425bba4dcf1231c256c6603750c8e1d/src/Avalonia.Controls/Primitives/AdornerLayer.cs#L177C38-L177C38) in `AdornerLayer.AddVisualAdorner` effectively prevents any adorner control from setting the attached `IsClipEnabled` property to true because that line seems to blow away any existing setting and executes after attached properties were already set in XAML.

It seems if the intent is for the default to be false, that line in `AddVisualAdorner` should be removed and the `IsClipEnabledProperty` default should be false instead.

## ScrollViewer or Other Container Clipping Issue

In our case, we want to render an adorner overlapping the adorned element but outside some of its rectangular element bounds.  Yet the adorner control should still be clipped by any container that would clip the adorned element.  Currently no clipping of ancestors takes place, which leads to the issues in the animation above and makes adorners almost unusable.  

I'm not sure what updates would be needed but `AdornerLayer.UpdateClip` should probably take whatever clip (or no clip) that comes out of it and intersect that with the bounds of the adorned element that are actually visible up through to the topmost visual.

This clip would need to be updated anytime the adorned element is arranged, possibly even when other ancestors are arranged to handle scenarios like the `ScrollViewer` one I showed.

**Comment by billhenn** [comment]
After the PR #13921 updates above, if I change the two `Canvas` controls that are adorners in the XAML above to be:
```
<Canvas AdornerLayer.IsClipEnabled="False">
```

The adorners still render outside of the `ScrollViewer`, which is bad, but the artifacts are no longer left behind like before.  Here's an animation showing what I see:

![AdornerIssue2](https://github.com/AvaloniaUI/Avalonia/assets/5760058/5b8d5c66-1901-4f2d-b890-720f3e40908a)

Any idea how we make sure that containers of the adorner target `Border` controls that would clip those `Border` controls also clip the adorners properly too?  Like in this scenario, the red/blue dots shouldn't be visible outside of the `ScrollViewer` content area.

**Comment by billhenn** [comment]
I would think that you have two general scenarios with adorners that you want to support...

1) You want them clipped to the same view/clip that the target element effectively has.  If the target element is partially scrolled off view, any part of the adorner that is also partially scrolled or clipped out of view by a container shouldn't be visible either.  That's the behavior I would expect with the default `AdornerLayer.IsClipEnabled="True"` since I think it's the most common scenario and is ideal for any kind of decoration.

Whereas right now with clip enabled and with the commits above, it's limiting the adorner rendering to only within the bounds of the target element and not outside.  This prevents scenarios like decoration dots/badges that extend beyond the target element's bounds from working properly, as they will be clipped like you see here with the blue dot's top half not being visible anymore:

![image](https://github.com/AvaloniaUI/Avalonia/assets/5760058/3511968b-6c86-49b1-9f7c-0a265028d065)

On a positive note, the `AdornerLayer.IsClipEnabled="True"` scenario is now working correctly with containing `ScrollViewer` controls.

2) When `AdornerLayer.IsClipEnabled="False"`, I would think that would be more of a special case like splitters where you don't care about clips at all and want it always visible.  That scenario seems to be working as expected.

To summarize, I think the thing we're missing is in the `AdornerLayer.IsClipEnabled="True"` scenario having the ability to render adorners outside of the bounds of the target element but still getting properly clipped by target element containers such as `ScrollViewer` controls.

**Comment by grokys** [comment]
> I think the thing we're missing is in the `AdornerLayer.IsClipEnabled="True"` scenario having the ability to render adorners outside of the bounds of the target element but still getting properly clipped by target element containers such as `ScrollViewer` controls.

WPF solves this problem by hosting an `AdornerLayer` in its `ScrollContentPresenter`:

![image](https://github.com/AvaloniaUI/Avalonia/assets/1775141/40371b4c-312e-4786-9a70-9cb1ad876343)

We could do the same, but I see various threads online of people wanting to disable this behavior.

**Comment by grokys** [comment]
While we decide what to do here, a workaround is to put a `VisualLayerManager` inside the `ScrollViewer` template:

```xml
    <ControlTheme x:Key="{x:Type ScrollViewer}" TargetType="ScrollViewer">
      <Setter Property="Template">
        <ControlTemplate>
          <Grid ColumnDefinitions="*,Auto"
                RowDefinitions="*,Auto">
            <VisualLayerManager>
              <ScrollContentPresenter Name="PART_ContentPresenter"
                                      Padding="{TemplateBinding Padding}"
                                      HorizontalSnapPointsType="{TemplateBinding HorizontalSnapPointsType}"
                                      VerticalSnapPointsType="{TemplateBinding VerticalSnapPointsType}"
                                      HorizontalSnapPointsAlignment="{TemplateBinding HorizontalSnapPointsAlignment}"
                                      VerticalSnapPointsAlignment="{TemplateBinding VerticalSnapPointsAlignment}"
                                      Background="{TemplateBinding Background}"
                                      ScrollViewer.IsScrollInertiaEnabled="{TemplateBinding IsScrollInertiaEnabled}">
                <ScrollContentPresenter.GestureRecognizers>
                  <ScrollGestureRecognizer CanHorizontallyScroll="{Binding CanHorizontallyScroll, ElementName=PART_ContentPresenter}"
                                           CanVerticallyScroll="{Binding CanVerticallyScroll, ElementName=PART_ContentPresenter}"
                                           IsScrollInertiaEnabled="{Binding (ScrollViewer.IsScrollInertiaEnabled), ElementName=PART_ContentPresenter}"/>
                </ScrollContentPresenter.GestureRecognizers>
              </ScrollContentPresenter>
            </VisualLayerManager>
            <ScrollBar Name="PART_HorizontalScrollBar"
                       Grid.Row="1"
                       Orientation="Horizontal"/>
            <ScrollBar Name="PART_VerticalScrollBar"
                       Grid.Column="1"
                       Orientation="Vertical"/>
            <Panel Grid.Row="1"
                   Grid.Column="1"
                   Background="{DynamicResource ThemeControlMidBrush}"/>
          </Grid>
        </ControlTemplate>
      </Setter>
    </ControlTheme>
```

With #13921 applied, this correctly clips adorners to the scroll viewer:

https://github.com/AvaloniaUI/Avalonia/assets/1775141/80c3e419-ccdc-4f6d-9f27-6303bef5d8ae

This isn't ideal however as:

- There's no way to enable/disable the behavior other than re-templating `ScrollViewer` (though that is an improvement as in WPF there's just no way to disable it)
- Anything that wants to use a visual layer manager from within the scroll viewer, such as overlap popups will now find the one inside the scrollviewer instead of the top-level one

@billhenn I'd suggest opening a new issue for this as it's technically a new feature rather than a bug, which is what the original issue was. I think we should close this issue when #13921 is merged.
